Libraries/StyleSheet/NativeColorValueTypes.ios.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @format
+ * @flow strict-local
+ */
+
+'use strict';
+
+import type {ColorValue, ProcessedColorValue} from './ColorValueTypes';
+
+export opaque type NativeColorValue = {
+  semantic?: Array<string>,
+  dynamic?: {
+    light: ?(ColorValue | ProcessedColorValue),
+    dark: ?(ColorValue | ProcessedColorValue),
+  },
+};
+
+export const PlatformColor = (...names: Array<string>): ColorValue => {
+  return {semantic: names};
+};
+
+export type IOSDynamicColorTuple = {
+  light: ColorValue,
+  dark: ColorValue,
+};
+
+export const IOSDynamicColor = (tuple: IOSDynamicColorTuple): ColorValue => {
```

Comment:
I think what you had before was reasonable and a common problem we have with platform extensions. 

If you have a project that is typechecked with different configs against both ios and android (as is the case for how we check facebook and react native code), and you have myfile.js with these contents:

```
{
  backgroundColor: Platform.OS === 'ios' ? IOSDynamicColor(...) : AndroidSpecialColor(...),
}
```

This file will fail to type check on both platforms. When type checked on iOS, `AndroidSpecialColor` won't type check, and when typechecked on Android, `IOSDynamicColor` won't exist.

I'm not sure how we should solve it in this instance, but we have often had to give up on platform extensions in those cases and just do `NativeColorValueTypesIOS.js` and `NativeColorValueTypesAndroid.js` and import the functions you want from both. 

I'm not sure how to make that work smoothly with the opaque type though...

Reply:
Oh, you're right.   I forgot that I do import `IOSDynamicColor` in the `PlatformColorExample.js` test and it's shared so now `yarn flow-check-android` is broken.

There could be a `NativeColorValueTypesIOS.ios.js` and `NativeColorValueTypesIOS.android.js` and so on with stubs in the other platforms.   But to overcome the `opaque` type problem there would have to be a level of indirection with platform specific "factory" methods that are only defined in the platform specific files.

I tried this locally and it works.   It will work fine in out-of-tree platforms too: they'll just have to have similar stub files: `NativeColorValueTypesIOS.macos.js`, `NativeColorValueTypesIOS.windows.js`, etc.

## NativeColorValueTypesIOS.ios.js:
```js
import type {ColorValue} from './ColorValueTypes';
import type {IOSDynamicColorTuplePrivate} from './NativeColorValueTypes';
import {IOSDynamicColorPrivate} from './NativeColorValueTypes';

export type IOSDynamicColorTuple = {
  light: ColorValue,
  dark: ColorValue,
};

export const IOSDynamicColor = (tuple: IOSDynamicColorTuple): ColorValue => {
  return IOSDynamicColorPrivate({light: tuple.light, dark: tuple.light});
};
```
## NativeColorValueTypesIOS.android.js:
```js
import type {ColorValue, ProcessedColorValue} from './ColorValueTypes';

export type IOSDynamicColorTuple = {
  light: ColorValue,
  dark: ColorValue,
};

export const IOSDynamicColor = (tuple: IOSDynamicColorTuple): ColorValue => {
  throw new Error("IOSDynamicColor is not available on this platform.");
};
```
## NativeColorValueTypes.ios.js:
```js
...

export type IOSDynamicColorTuplePrivate = {
  light: ColorValue,
  dark: ColorValue,
};

export const IOSDynamicColorPrivate = (tuple: IOSDynamicColorTuplePrivate): ColorValue => {
  return {dynamic: {light: tuple.light, dark: tuple.dark}};
};

...
```

Thoughts?

Reply:
The downside of this is that every platform has to know of every other platform to be able to include stubs for that. I'd like to avoid a situation where the maintainers of MyNewRokuPlatform need to go send PRs to every other platform to add stubs for their new platform color.

I think I'll need to talk to the Flow team to see if they have any recommendations.

Reply:
Mmm, but the stubs can actually just be in `NativeColorValueTypesIOS.js` (no platform extension).   Then the out-of-tree platforms don't need to create their own stubs.

But the reverse problem of an out-of-tree platform specific method (e.g. `MacOSDynamicColor()`) can't be used in an App's .js file that was intended to be packaged for both in-tree *and* out-of-tree platforms.   We have this problem now and our our solutions are for the App to define their colors into platform specific files like:
`Colors.common.ts`
`Colors.ios.ts`
`Colors.android.ts`
`Colors.macos.ts`
`Colors.windows.ts`
The `Colors.common.ts` defines a dictionary of App color keys and each value is magenta.
Each `Colors.<platform>.ts` file imports `Colors.common.ts`, spreads the common App colors and then replaces each one using whatever platform specific method is appropriate.   If a new color is added and a platform file forgets to override it its obvious during testing because it appears as magenta.

## SomeApp/src/styles/Colors.common.ts
```ts
var UNDEFINED_COLOR = "#ff00ff";
export default {
    someColor: UNDEFINED_COLOR,
. . .
```
## SomeApp/src/styles/Colors.macos.ts
```ts
import Common from "./Colors.common";
import {MacDynamicColor} from "NativeColorValueTypesMacOS";
export default {
  ...Common,
  someColor: MacDynamicColor({light: "black", dark: "white"}),
. . .
```

Reply:
Yeah, having to do that split and have a file with all the colors doesn't feel great. I'd like people to be able to just do:

```
// App.js
const styles = StyleSheet.create({
  root: {
    backgroundColor: Platform.OS === 'ios' ? IOSDynamicColor(...) : MacDynamicColor(...),
  }
})
```

Reply:
Actually, with proper metro and flow config I think this pattern would work for both in-tree and out-of-tree platforms given that a platform repo provides both its own platform specific .js file *and* the platform agnostic stub .js file just as core would for ios and android.

So in core there would be:
```
Libraries/StyleSheet/NativeColorValueTypesIOS.ios.js  # contains real IOSDynamicColor()
Libraries/StyleSheet/NativeColorValueTypesIOS.js      # contains stub IOSDynamicColor()
```
and in another platform repo there could be:
```
Libraries/StyleSheet/NativeColorValueTypesRoku.roku.js  # contains real RokuColor()
Libraries/StyleSheet/NativeColorValueTypesRoku.js       # contains stub RokuColor()
```
Then an App repo that bundles for both iOS and Roku could in a .js file import both `IOSDynamicColor` and `RokuColor` and do:
```
// App.js
const styles = StyleSheet.create({
  root: {
    backgroundColor: Platform.OS === 'roku' ? RokuColor(...) : IOSDynamicColor(...),
  }
})
```

Reply:
Ooh. Interesting. And then our Flow config and metro config internally would ignore the unprefixed file? 

Okay, I could see that making sense. 

Any ideas on how we can get the opaque type working with this, or at least make it sufficiently painful for people to not call these methods directly?

Reply:
I think I did make it sufficiently painful 😄 .   The only ways to create an opaque `ColorValue` that is an iOS dynamic color is to `import {IOSDynamicColor} from 'NativeColorValueTypesIOS'` and call `IOSDynamicColor()` *or*  `import {IOSDynamicColorPrivate} from 'NativeColorValueTypes.ios'` and call `IOSDynamicColorPrivate()`.   If someone did call `IOSDynamicColorPrivate()` it would work on iOS but throw in the native layer of non-IOS platforms.

Reply:
Got it, so the approach isn't to use an opaque type, is that correct?

In your current changes you have 
```
import type {NativeColorValue} from './NativeColorValueTypes';

export type ColorValue = null | string | NativeColorValue;
```

Since `NativeColorValueTypes` uses platform extensions each defining that opaque type, it seems like the previous example would fail to typecheck against the ios flowconfig, since RokuColor would return the opaque RokuColor type, whereas the ios flow config wants them to all match the opaque iOS DynamicColor type. 

```
const styles = StyleSheet.create({
  root: {
    backgroundColor: Platform.OS === 'roku' ? RokuColor(...) : IOSDynamicColor(...),
  }
})
```

I've never used opaque types in combination with platform extension files, so it is possible that this somehow magically works in Flow, but my intuition is that it wouldn't.

Can you add an example like the above roku one but for ios/android to the RNTesterExample? That should make sure our solution handles this situation long term.

Reply:
It totally works!   `yarn flow-check-ios ; yarn flow-check-android` passes and bundling works for both plats:

![image](https://user-images.githubusercontent.com/30053638/73982069-6c9c0c00-48e8-11ea-96fc-34974ceb85b5.png)

Its quite satisfying.    See the changes in `PlatformColorsExample.js`.

Reply:
I don't see an example that's quite what I'm looking for, although I think the behavior might be handled by parts of other tests. Can you add an explicit case that is like:

```
function VariantColorsExample() {
  return (
    <View style={styles.column}>
      <View style={styles.row}>
        <Text style={styles.labelCell}>
        {
          Platform.OS === 'ios' ? "IOSDynamicColor({light: 'red', dark: 'blue'})" : "AndroidColor('?attr/colorAccent')",
        }
        </Text>
        <View
          style={{
            ...styles.colorCell,
            backgroundColor: Platform.OS === 'ios' ? IOSDynamicColor({light: 'red', dark: 'blue'}) : AndroidColor('?attr/colorAccent'),
          }}
        />
      </View>
    </View>
  );
}
```

Reply:
No problem.   Pushed a change to add a `VariantColorsExample` as above.

![image](https://user-images.githubusercontent.com/30053638/74108801-98123700-4b32-11ea-9dc3-b4b6337fa384.png)